### PR TITLE
[DCOS-59796]: Integration test for Operator tenancy

### PR DESCRIPTION
**Pull request by alembiewski**
### What changes were proposed in this pull request?
Resolves [DCOS-59796](https://jira.mesosphere.com/browse/DCOS-59796)

### Why are the changes needed?
Add a set of integration tests to verify operator multi-tenancy by checking the main components (`Service`, `ServiceAccount`, `Role`, `ClusterRole`, `ClusterRoleBinding`, `CRD`) and by running a simple workload on each installation.

Tests cover the following scenarios:
- install two operator instances in two namespaces
- install two operator instances with the same name in two namespaces
- install two operator instances in a single namespace
- CRDs deletion

### How were the changes tested?
tested locally on `minikube` and by running the build in CI

### CODE REVIEW

File: tests/templates/spark-pi.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+apiVersion: "sparkoperator.k8s.io/v1beta2"
```
Comment (akirillov) [inline tests/templates/spark-pi.yaml]: can we use mock task runner job as default here? nothing prevents us from using Spark Pi but MTR is deterministic with respect to execution time and can be launched with some small sleep time to avoid long waits.

Comment (alembiewski) [reply tests/templates/spark-pi.yaml]: good point, but I would suggest to keep using `SparkPi` as a test job because it is faster in comparison with `MockTaskRunner`. In case with the latter, a job needs to fetch an external artifact (> 60 Mb), and that will increase test times, and for the former, the jar is already present inside the image. I've compared two examples locally, `SparkPi` was a way faster then MTR, started with the small arguments' values (1 task 1 second).

Comment (akirillov) [reply tests/templates/spark-pi.yaml]: good point, thanks. we download MTR anyway for testing the killing of running application where we need more control over the duration to avoid flakiness due to timing issues - so the download time is already there.

We also user Spark Pi in DCOS Spark tests to verify Driver output - so leaving it up to you, just wanted to keep the number of simple apps (not targeting any specific corner-cases like shuffle) low.

Comment (alembiewski) [reply tests/templates/spark-pi.yaml]: > we download MTR anyway for testing the killing of running application where we need more control over the duration to avoid flakiness due to timing issues - so the download time is already there.

yes, but from my observations during local testing, it won't be reused across different operator instances even in the same namespace from one test run to another.

**Comment by alembiewski** [comment]
Thanks for the review @akirillov! 
CI check has passed, but I would like to postpone merging this PR before #45 is merged because both PRs contain similar changes related to conditional cleanup in `InstallSparkOperator` method, and I like @rpalaznik's approach of handling that, so I'm going to rebase my branch on top of this changes.

File: tests/tenancy_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		for _, role := range roles {
+			if strings.Contains(role, "%s") {
+				role = fmt.Sprintf(role, operator.InstanceName)
+			}
+			log.Infof("Checking Role \"%s\" is created in namespace \"%s\" for \"%s\"",
+				role, operator.Namespace, operator.InstanceName)
+			result, err := operator.K8sClients.RbacV1().Roles(operator.Namespace).Get(role, v1.GetOptions{})
+			assert.NilError(t, err)
+			instanceLabel, present := result.Labels["kudo.dev/instance"]
+			if present {
```
Comment (VladimirKravtsov) [inline tests/tenancy_test.go]: Why if the label is not present here, the situation is ignored?

File: tests/tenancy_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+}
+
+func verifyComponents(t *testing.T, operators []*utils.SparkOperatorInstallation) {
+	serviceAccounts := []string{"spark-operator-service-account", "spark-service-account"}
+	services := []string{"spark-webhook", "spark-operator-metrics"}
+	roles := []string{"spark-driver-role", "%s-spark-role"}
+
+	for _, operator := range operators {
+		for _, service := range services {
```
Comment (VladimirKravtsov) [inline tests/tenancy_test.go]: I'd suggest wrapping each following subtests (verify Services, ServiceAccounts, Roles etc.) into `t.Run`, so making them atomical. It will more comply with the unit testing philosophy.

The example of how it could looks like
```suggestion
for _, operator := range operators {
        t.Run("TestServices", func(t *testing.T) {
		for _, service := range services {
					serviceName := fmt.Sprint(operator.InstanceName, "-", service)
 			log.Infof("Checking Service \"%s\" is created in namespace \"%s\" for \"%s\"", serviceName,
 				operator.Namespace, operator.InstanceName)
 			result, err := operator.K8sClients.CoreV1().Services(operator.Namespace).Get(fmt.Sprint(serviceName), v1.GetOptions{})
 			assert.NilError(t, err)
 			assert.Equal(t, result.Labels["kudo.dev/instance"], operator.InstanceName)
 		}
 	}(t)

```

File: tests/tenancy_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+func TestTenancyTwoOperatorsSameNameDifferentNamespaces(t *testing.T) {
+	operators := operatorBuilder(2, true, false)
+	for _, operator := range operators {
+		err := operator.InstallSparkOperator(true)
+		assert.NilError(t, err)
+	}
+
+	verifyComponents(t, operators)
+	verifyWorkloads(t, operators)
```
Comment (VladimirKravtsov) [inline tests/tenancy_test.go]: Here and in the above method:
```suggestion
t.Run("TestWorkloads", func(t *testing.T) {
	verifyWorkloads(t, operators)
})
```

File: tests/tenancy_test.go (hunk 3 above)
Comment (VladimirKravtsov) [reply tests/tenancy_test.go]: Also, the tests which are wrapped into `t.Run` could be run in parallel. The optimization can bring faster execution and atomical the tests fails.

File: tests/tenancy_test.go (hunk 2 above)
Comment (alembiewski) [reply tests/tenancy_test.go]: yes, because in this particular case not all the `Role` resources contain this label that we are verifying, therefore we can't explicitly check to which instance it belongs to. But we are still checking the role exists it the operator's namespace, so this test is valid.

File: tests/templates/spark-pi.yaml (hunk 1 above)
Comment (akirillov) [reply tests/templates/spark-pi.yaml]: that's right. Spark pulls application jar every time a job is submitted, good point.

**Comment by alembiewski** [comment]
Thanks for the review, @VladimirKravtsov!
I decomposed the tests based on the approach you suggested,  and it looks good from my perspective: tests became more granular and self-descriptive. Here is how it looks like in CI:
https://teamcity.mesosphere.io/viewLog.html?buildTypeId=Frameworks_DataServices_Kudo_Spark_Pr_SparkPrKonvoyKudoV073&buildId=2400816&tab=testsInfo&pager.currentPage=1

File: tests/tenancy_test.go (hunk 2 above)
Comment (VladimirKravtsov) [reply tests/tenancy_test.go]: ok, got it

**Comment by akirillov** [comment]
just as a side-note, I think with the new `t.Run` approach the test cases reports became fuzzier. The semantics of the test cases here is the following: "All the components of the Operator are installed under these conditions". In the test report, it breaks down to 6 independent tests per each condition which makes it hard to reason about what was the main scenario.

CI now reports 45 independent tests which I believe does not reflect the actual use cases. I'm not sure it's worth the parallel optimization:
CI test report sample

**Comment by alembiewski** [comment]
@akirillov, 
I agree, that we probably don't need such detailed report about all of the components, so I've made the test-suite less granular:
image
